src/Init/NotationExtra.lean

```diff
@@ -254,7 +254,7 @@ macro_rules
 syntax "repeat " doSeq " until " term : doElem
 
 macro_rules
-  | `(doElem| repeat $seq until $cond) => `(doElem| repeat do $seq; if $cond then break)
+  | `(doElem| repeat $seq until $cond) => `(doElem| repeat do $seq:doSeq; if $cond then break)
```

Comment:
I think this one makes sense: how many people would read the original line and conclude that `$seq` is not the first of two elements of `do`, but the entire content of it? With the relaxed antiquotation parser, we do now get ``seq : TSyntax `doElem`` in the modified line.
The fix only makes this marginally clearer; I would have liked `(do $seq); ...` better, which does interpret `seq` as a `doSeq`. However, the parens terminate e.g. mutable variables, which is undesirable.

Reply:
To be fair, I can empathize with Lean here.  I've also found the zoo of do-notation nonterminals a bit confusing at times.

Indentation would be another way to disambiguate this example:
```lean
`(doElem|
  repeat
    do $seq
    if $cond then break)
```

Excessive use of braces helps too:
```lean
`(doElem| repeat do { do $seq }; if $cond then break)
```

We could also remove the slightly misleading syntax for nested do and replace it by braces.  (I think it's misleading because it's not a boundary for `return` like the normal `do`.)
```lean
syntax "{" doSeq "}" : doElem
```

A coercion from `doSeq` to `doSeqItem` could also help (but I guess you don't want a coercion to insert new tokens).

Reply:
Yeah, it might be good to clean up the grammar there. And yes introducing tokens in a coercion probably isn't a good idea due to lack of access to `getRef` (though I even thought about a monadic coercion class that is automatically used by antiquotations).